tests/test_integration.py

```diff
@@ -0,0 +1,137 @@
+        stripe.api_key = "sk_test_123"
+        stripe.default_http_client = None
+        stripe.log = "warning"
+        stripe.proxy = None
+        yield
+        stripe.api_base = orig_attrs["api_base"]
+        stripe.api_key = orig_attrs["api_key"]
+        stripe.default_http_client = orig_attrs["default_http_client"]
+        stripe.log = orig_attrs["log"]
+        stripe.proxy = orig_attrs["proxy"]
+
+    def setup_mock_server(self, handler):
+        # Configure mock server.
+        self.mock_server_port = get_free_port()
+        self.mock_server = HTTPServer(
+            ("localhost", self.mock_server_port), handler
+        )
+
+        # Start running mock server in a separate thread.
+        # Daemon threads automatically shut down when the main process exits.
+        self.mock_server_thread = Thread(target=self.mock_server.serve_forever)
+        self.mock_server_thread.setDaemon(True)
+        self.mock_server_thread.start()
+
+    def test_hits_api_base(self):
+        class MockServerRequestHandler(BaseHTTPRequestHandler):
+            num_requests = 0
+
+            def do_GET(self):
+                self.__class__.num_requests += 1
+
+                self.send_response(200)
+                self.send_header(
+                    "Content-Type", "application/json; charset=utf-8"
+                )
+                self.end_headers()
+                self.wfile.write(json.dumps({}).encode("utf-8"))
+                return
+
+        self.setup_mock_server(MockServerRequestHandler)
+
+        stripe.api_base = "http://localhost:%s" % self.mock_server_port
+        stripe.Balance.retrieve()
+        assert MockServerRequestHandler.num_requests == 1
+
+    def test_hits_stripe_proxy(self, mocker):
+        class MockServerRequestHandler(BaseHTTPRequestHandler):
+            num_requests = 0
+
+            def do_GET(self):
+                self.__class__.num_requests += 1
+
+                self.send_response(200)
+                self.send_header(
+                    "Content-Type", "application/json; charset=utf-8"
+                )
+                self.end_headers()
+                self.wfile.write(json.dumps({}).encode("utf-8"))
+                return
+
+        self.setup_mock_server(MockServerRequestHandler)
+        logger_mock = mocker.patch("stripe.util.log_warning")
+
+        stripe.proxy = "http://localhost:%s" % self.mock_server_port
+        stripe.Balance.retrieve()
+        assert MockServerRequestHandler.num_requests == 1
+
+        stripe.proxy = "http://bad-url"
+        logger_mock.assert_not_called()
+        stripe.Balance.retrieve()
+        logger_mock.assert_called_with(
+            "stripe.proxy was updated after sending a request - this is a no-op. To use a different proxy, set stripe.default_http_client to a new client configured with the proxy."
+        )
+        assert MockServerRequestHandler.num_requests == 2
+
+    def test_hits_client_proxy(self):
```

Comment:
Can we name this one something like `test_hits_proxy_through_default_http_client`?